### PR TITLE
monitoring: reduce src_graphql_field_seconds cardinality substantially

### LINKED ISSUE #9895: reduce src_graphql_field_seconds cardinality
Right now the series has 2500+ unique pairs, which is crazy high cardinality. Most of it comes from the `field` value having such high cardinality. A complete list of these values from Sourcegraph.com:

```
   2 AccessTokenConnection nodes
   2 AccessTokenConnection pageInfo
   2 AccessTokenConnection totalCount
   6 AuthProvider authenticationURL
   6 AuthProvider clientID
   6 AuthProvider displayName
   6 AuthProvider isBuiltin
   6 AuthProvider serviceID
   6 AuthProvider serviceType
   6 AuthProviderConnection nodes
   6 ClientConfigurationDetails contentScriptUrls
   6 ClientConfigurationDetails parentSourcegraph
   1 CommitSearchResult commit
   6 CommitSearchResult detail
   1 CommitSearchResult diffPreview
   6 CommitSearchResult icon
   6 CommitSearchResult label
   6 CommitSearchResult matches
   1 CommitSearchResult messagePreview
   1 CommitSearchResult refs
   1 CommitSearchResult sourceRefs
   6 CommitSearchResult url
   6 Configuration contents
   6 Configuration messages
   6 ConfigurationCascade merged
   6 ConfigurationCascade subjects
   6 DiffStat added
   6 DiffStat changed
   6 DiffStat deleted
   2 DiscussionThreadConnection nodes
   2 DiscussionThreadConnection pageInfo
   2 DiscussionThreadConnection totalCount
   1 DotcomQuery previewProductSubscriptionInvoice
   2 DotcomQuery productPlans
   4 DotcomQuery productSubscriptions
   6 ExtensionManifest description
   6 ExtensionManifest raw
   4 ExtensionRegistry extension
   6 ExtensionRegistry extensions
   6 ExternalLink serviceType
   6 ExternalLink url
   6 File isDirectory
   6 File name
   6 File path
   6 File repository
   6 File url
   6 File2 content
   6 File2 externalURLs
  12 File2 highlight
   6 File2 isDirectory
   6 File2 richHTML
   6 File2 url
   6 FileDiff hunks
   6 FileDiff internalID
   6 FileDiff mostRelevantFile
   6 FileDiff newPath
   6 FileDiff oldPath
   6 FileDiff stat
   6 FileDiffConnection diffStat
   6 FileDiffConnection nodes
   6 FileDiffConnection pageInfo
   6 FileDiffConnection totalCount
   6 FileDiffHunk body
   6 FileDiffHunk newRange
   6 FileDiffHunk oldNoNewlineAt
   6 FileDiffHunk oldRange
   6 FileDiffHunk section
   6 FileDiffHunkRange lines
   6 FileDiffHunkRange startLine
   6 FileMatch file
   6 FileMatch limitHit
   6 FileMatch lineMatches
   6 FileMatch repository
   6 FileMatch revSpec
   6 FileMatch symbols
   6 GitBlob blame
   6 GitBlob commit
   5 GitBlob content
   7 GitBlob lsif
   6 GitBlob path
   6 GitBlob repository
   6 GitBlob url
   6 GitCommit abbreviatedOID
   6 GitCommit ancestors
   6 GitCommit author
  12 GitCommit blob
   6 GitCommit body
   6 GitCommit canonicalURL
   6 GitCommit committer
   6 GitCommit externalURLs
  12 GitCommit file
   6 GitCommit id
   6 GitCommit message
   6 GitCommit oid
   6 GitCommit parents
   1 GitCommit repository
   6 GitCommit subject
   6 GitCommit symbols
   7 GitCommit tree
   6 GitCommit url
   6 GitCommitConnection nodes
   6 GitCommitConnection pageInfo
   5 GitObject commit
   2 GitObject oid
   6 GitRef abbrevName
   6 GitRef displayName
   2 GitRef id
   3 GitRef name
   1 GitRef prefix
   1 GitRef repository
   2 GitRef target
   2 GitRef url
   2 GitRefConnection nodes
   2 GitRefConnection pageInfo
   2 GitRefConnection totalCount
   5 GitRevSpecExpr expr
   5 GitRevSpecExpr object
   2 GitRevisionRange baseRevSpec
   2 GitRevisionRange expr
   2 GitRevisionRange headRevSpec
   6 GitTree canonicalURL
   6 GitTree entries
   6 GitTree files
   6 GitTree isRoot
   6 GitTree url
   1 Highlight character
   1 Highlight length
   1 Highlight line
   6 HighlightedFile aborted
   6 HighlightedFile html
   1 HighlightedString highlights
   1 HighlightedString value
   6 Hover markdown
   6 Hover range
   6 Hunk author
   6 Hunk commit
   6 Hunk endLine
   6 Hunk message
   6 Hunk rev
   6 Hunk startLine
   6 LSIFQueryResolver definitions
   6 LSIFQueryResolver hover
   6 LineMatch lineNumber
   6 LineMatch offsetAndLengths
   6 LineMatch preview
   6 Location range
   6 Location resource
   6 LocationConnection nodes
   6 Markdown html
   6 Markdown text
   6 MirrorRepositoryInfo cloneInProgress
   6 MirrorRepositoryInfo cloneProgress
   6 MirrorRepositoryInfo cloned
   5 Mutation configurationMutation
   2 Mutation createOrganization
   6 Mutation logEvent
   6 Mutation logUserEvent
   1 Org createdAt
   6 Org displayName
   6 Org id
   2 Org members
   6 Org name
   4 Org settingsURL
   6 Org url
   3 Org viewerCanAdminister
   1 Org viewerIsMember
   1 Org viewerPendingInvitation
   6 OrgConnection nodes
   6 PageInfo hasNextPage
   6 ParentSourcegraphDetails url
   6 Person avatarURL
   6 Person displayName
   6 Person email
   6 Person name
   6 Person user
   6 Position character
   6 Position line
   6 ProductLicenseInfo expiresAt
   2 ProductPlan billingPlanID
   2 ProductPlan minQuantity
   2 ProductPlan name
   2 ProductPlan planTiers
   2 ProductPlan pricePerUserPerYear
   2 ProductPlan productPlanID
   2 ProductPlan tiersMode
   4 ProductSubscriptionConnection nodes
   4 ProductSubscriptionConnection pageInfo
   4 ProductSubscriptionConnection totalCount
   1 ProductSubscriptionInvoiceItem expiresAt
   1 ProductSubscriptionInvoiceItem plan
   1 ProductSubscriptionInvoiceItem userCount
   1 ProductSubscriptionPreviewInvoice afterInvoiceItem
   1 ProductSubscriptionPreviewInvoice beforeInvoiceItem
   1 ProductSubscriptionPreviewInvoice isDowngradeRequiringManualIntervention
   1 ProductSubscriptionPreviewInvoice price
   1 ProductSubscriptionPreviewInvoice prorationDate
   6 ProductSubscriptionStatus license
   6 ProductSubscriptionStatus noLicenseWarningUserCount
   6 Query clientConfiguration
   6 Query currentUser
   2 Query discussionThreads
   5 Query dotcom
   6 Query extensionRegistry
   6 Query highlightCode
   7 Query node
   1 Query organization
   4 Query repositories
  12 Query repository
   9 Query repositoryRedirect
   6 Query search
   6 Query settingsSubject
   6 Query site
   5 Query user
   6 Query viewerConfiguration
   6 Query viewerSettings
   6 Range end
   6 Range start
   6 RegistryExtension createdAt
   6 RegistryExtension extensionID
   6 RegistryExtension extensionIDWithoutRegistry
   6 RegistryExtension id
   6 RegistryExtension isLocal
   6 RegistryExtension isWorkInProgress
   6 RegistryExtension manifest
   6 RegistryExtension name
   4 RegistryExtension publishedAt
   6 RegistryExtension publisher
   6 RegistryExtension registryName
   6 RegistryExtension remoteURL
   6 RegistryExtension updatedAt
   6 RegistryExtension url
   6 RegistryExtension viewerCanAdminister
   5 RegistryExtensionConnection error
   6 RegistryExtensionConnection nodes
   4 Repository cloneInProgress
  12 Repository commit
   6 Repository comparison
   6 Repository defaultBranch
   6 Repository description
   6 Repository detail
   6 Repository externalURLs
   2 Repository gitRefs
   6 Repository icon
   6 Repository id
   6 Repository isArchived
   6 Repository isFork
   6 Repository label
   6 Repository matches
   6 Repository mirrorInfo
   6 Repository name
   5 Repository redirectURL
   6 Repository uri
   6 Repository url
   6 Repository viewerCanAdminister
   3 RepositoryComparison commits
   6 RepositoryComparison fileDiffs
   2 RepositoryComparison range
   4 RepositoryConnection nodes
   4 RepositoryConnection pageInfo
   4 RepositoryConnection totalCount
  11 Search results
  12 Search suggestions
   6 SearchAlert description
   6 SearchAlert proposedQueries
   6 SearchAlert title
   6 SearchFilter count
   6 SearchFilter kind
   6 SearchFilter label
   6 SearchFilter limitHit
   6 SearchFilter value
   6 SearchQueryDescription description
   6 SearchQueryDescription query
   6 SearchResultMatch body
   6 SearchResultMatch highlights
   6 SearchResultMatch url
  10 SearchResults alert
  10 SearchResults approximateResultCount
  10 SearchResults cloning
   6 SearchResults dynamicFilters
   6 SearchResults elapsedMilliseconds
   6 SearchResults indexUnavailable
  10 SearchResults limitHit
   6 SearchResults matchCount
   6 SearchResults missing
   6 SearchResults repositoriesCount
  10 SearchResults results
  10 SearchResults timedout
   6 Session canSignOut
   6 Settings configuration
   6 Settings contents
   6 Settings id
   6 SettingsCascade final
   6 SettingsCascade subjects
   9 SettingsMutation editConfiguration
   6 SettingsSubject latestSettings
   6 SettingsSubject settingsCascade
   6 SettingsSubject settingsURL
   6 SettingsSubject viewerCanAdminister
   6 Signature date
   6 Signature person
   6 Site alerts
   6 Site authProviders
   6 Site buildVersion
   6 Site disableBuiltInSearches
   6 Site freeUsersExceeded
   6 Site hasCodeIntelligence
   6 Site id
   6 Site needsRepositoryConfiguration
   6 Site productSubscription
   6 Site productVersion
   6 Site sendsEmailVerificationEmails
   6 Site siteID
  10 Site updateCheck
   3 Submodule commit
   3 Submodule url
   6 Symbol containerName
   6 Symbol fileLocal
   6 Symbol kind
   6 Symbol language
   6 Symbol location
   6 Symbol name
   6 Symbol url
   6 SymbolConnection nodes
   6 SymbolConnection pageInfo
   6 TreeEntry isDirectory
   6 TreeEntry isSingleChild
   6 TreeEntry name
   6 TreeEntry path
   6 TreeEntry submodule
   6 TreeEntry url
   4 UpdateCheck checkedAt
   4 UpdateCheck errorMessage
   4 UpdateCheck pending
   4 UpdateCheck updateVersionAvailable
   5 UpdateSettingsPayload empty
   2 User accessTokens
   6 User avatarURL
   5 User builtinAuth
   5 User createdAt
   6 User databaseID
   6 User displayName
   6 User email
   6 User emails
   6 User id
   6 User organizations
   6 User session
   6 User settingsURL
   6 User siteAdmin
   6 User tags
   6 User url
   6 User username
   6 User viewerCanAdminister
   4 User viewerCanChangeUsername
   2 UserConnection nodes
   2 UserConnection totalCount
   6 UserEmail email
   3 UserEmail isPrimary
   3 UserEmail verificationPending
   6 UserEmail verified
   3 UserEmail viewerCanManuallyVerify
   4 __Directive args
   4 __Directive description
   4 __Directive locations
   4 __Directive name
   4 __EnumValue deprecationReason
   4 __EnumValue description
   4 __EnumValue isDeprecated
   4 __EnumValue name
   4 __Field args
   4 __Field deprecationReason
   4 __Field description
   4 __Field isDeprecated
   6 __Field name
   4 __Field type
   4 __InputValue defaultValue
   4 __InputValue description
   4 __InputValue name
   4 __InputValue type
   4 __Schema directives
   4 __Schema mutationType
   4 __Schema queryType
   4 __Schema subscriptionType
   4 __Schema types
   4 __Type description
   4 __Type enumValues
   6 __Type fields
   4 __Type inputFields
   4 __Type interfaces
   4 __Type kind
   4 __Type name
   4 __Type ofType
   4 __Type possibleTypes
   4 null __schema
   6 null __type
   6 null __typename
```

### PR DESCRIPTION
**Pull request by slimsag**
Reduces the cardinality of `src_graphql_field_seconds` substantially by using a whitelist of `(type, field)` pairs. This initial whitelist is based on my own educated guesses, we can make the logic more robust in the future.

Fixes #9895

### CODE REVIEW

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 1]
```diff
@@ -115,6 +115,161 @@ func (prometheusTracer) TraceField(ctx context.Context, label, typeName, fieldNa
 	}
 }
 
+// prometheusFieldName reduces the cardinality of GraphQL field names to make it suitable
+// for use in a Prometheus metric. We only track the ones most valuable to us.
+//
+// See https://github.com/sourcegraph/sourcegraph/issues/9895
+func prometheusFieldName(typeName, fieldName string) string {
+	whitelisted := [][2]string{
```
Comment (uwedeportivo) [inline cmd/frontend/graphqlbackend/graphqlbackend.go:123]: can you please take this out in a var outside the function body

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 2]
```diff
@@ -115,6 +115,161 @@ func (prometheusTracer) TraceField(ctx context.Context, label, typeName, fieldNa
...
+		{"SymbolConnection", "nodes"},
+		{"TreeEntry", "isDirectory"},
+		{"TreeEntry", "isSingleChild"},
+		{"TreeEntry", "name"},
+		{"TreeEntry", "path"},
+		{"TreeEntry", "submodule"},
+		{"TreeEntry", "url"},
+		{"UserConnection", "nodes"},
+	}
+	for _, pair := range whitelisted {
```
Comment (uwedeportivo) [inline cmd/frontend/graphqlbackend/graphqlbackend.go:264]: the list seems sorted. do a binary search on it ? it's relatively small but still, this gets hit often enough

**Comment by slimsag** [comment]
not sure the easiest way to do a binary search on it and I don't have a ton of time, so instead I opted for a benchmark which shows it is under a few hundred ns/op:

```
$ go test -v -run=XXX -bench=Prometheus ./cmd/frontend/graphqlbackend/
goos: darwin
goarch: amd64
pkg: github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend
BenchmarkPrometheusFieldName
BenchmarkPrometheusFieldName/test-0
BenchmarkPrometheusFieldName/test-0-16         	 5128797	       234 ns/op
BenchmarkPrometheusFieldName/test-1
BenchmarkPrometheusFieldName/test-1-16         	 4854226	       244 ns/op
BenchmarkPrometheusFieldName/test-2
BenchmarkPrometheusFieldName/test-2-16         	 3212312	       367 ns/op
PASS
ok  	github.com/sourcegraph/sourcegraph/cmd/frontend/graphqlbackend	4.754s
```

**Comment by uwedeportivo** [comment]
👍

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 3]
```diff
@@ -115,6 +115,162 @@ func (prometheusTracer) TraceField(ctx context.Context, label, typeName, fieldNa
 	}
 }
 
+var whitelistedPrometheusFieldNames = [][2]string{
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/graphqlbackend.go:118]: qq: how this list is generated/gathered?

**Comment by keegancsmith** [comment]
> +var whitelistedPrometheusFieldNames = [][2]string{
>
> qq: how this list is generated/gathered?

It is a whitelist of fields we want to track. It is not generated, we
explicitly opt-in to collecting a field. The names come from the graphql
schema. If you want to monitor in prometheus how long a certain graphql
field takes, you need to add it to here now.

**Comment by unknwon** [comment]
@keegancsmith 
> +var whitelistedPrometheusFieldNames = [][2]string{ qq: how this list is generated/gathered?
> It is a whitelist of fields we want to track. It is not generated, we explicitly opt-in to collecting a field. The names come from the graphql schema. If you want to monitor in prometheus how long a certain graphql field takes, you need to add it to here now.

Thanks! Was just wondering who is on the other side of the mapping.

**Pull request merged**